skfp/distances/russell.py

```diff
@@ -171,3 +172,168 @@ def russell_binary_distance(
     0.5
     """
     return 1 - russell_binary_similarity(vec_a, vec_b)
+
+
+@validate_params(
+    {"X": ["array-like"], "Y": ["array-like", None]},
+    prefer_skip_nested_validation=True,
+)
+def bulk_russell_binary_similarity(
+    X: np.ndarray,
+    Y: Optional[np.ndarray] = None,
+) -> np.ndarray:
+    r"""
+    Bulk Russell similarity for binary matrices.
+
+    Computes the pairwise Russell similarity between binary matrices. If one array is
+    passed, similarities are computed between its rows. For two arrays, similarities
+    are between their respective rows, with `i`-th row and `j`-th column in output
+    corresponding to `i`-th row from first array and `j`-th row from second array.
+
+    See also :py:func:`russell_binary_similarity`.
+
+    Parameters
+    ----------
+    X : ndarray
+        First binary input array, of shape :math:`m \times m`
+
+    Y : ndarray, default=None
+        Second binary input array, of shape :math:`n \times n`. If not passed, similarities
+        are computed between rows of X.
+
+    Returns
+    -------
+    similarities : ndarray
+        Array with pairwise Russell similarity values. Shape is :math:`m \times n` if two
+        arrays are passed, or :math:`m \times m` otherwise.
+
+    See Also
+    --------
+    :py:func:`russell_binary_similarity` : Russell similarity function for two vectors.
+
+    Examples
+    --------
+    >>> from skfp.distances import bulk_russell_binary_similarity
+    >>> import numpy as np
+    >>> X = np.array([[1, 1, 1], [0, 0, 1]])
+    >>> Y = np.array([[1, 0, 1], [0, 1, 1]])
+    >>> sim = bulk_russell_binary_similarity(X, Y)
+    >>> sim
+    array([[0.66666667, 0.66666667],
+           [0.33333333, 0.33333333]])
+    """
+    if Y is None:
+        return _bulk_russell_binary_similarity_single(X)
+    else:
+        return _bulk_russell_binary_similarity_two(X, Y)
+
+
+@numba.njit(parallel=True)
+def _bulk_russell_binary_similarity_single(
+    X: np.ndarray,
+) -> np.ndarray:
+    m = X.shape[0]
+    sims = np.empty((m, m))
+
+    # upper triangle - actual similarities
+    for i in numba.prange(m):
+        vec_a = X[i]
+        len_a = len(vec_a)
```

Comment:
Length is constant for the entire matrix, it's `X.shape[1]`. You can bring it before the whole loop